### DISCUSSION TITLE
Check paddle billing webhook

### DISCUSSION THREAD

**Discussion by lonhtm** [pay-rails/pay#922] [Q&A] [answered]
I have been playing with pay and paddle billing for a week. Till now, I can now have a inline checkout in a turbo_frame. But all the subcriptions create by using paddle.js is not sync with the database. I think the webhook might not be properly config. Can you guy advices where should I have to dig deep to solve the issue? 

Thank you

**Comment by fidalgo** [comment]
I've been facing the same issue and while I'm debugging I will try to update here, so we can collectively find out the reason.
In my case, my model is Team and when I create one new team, my `Pay::Customer` data does not have the processor_id:
```
Pay::Customer:0x00007f8913a3f428
 id: 3,
 owner_type: "Team",
 owner_id: 3,
 processor: "paddle_billing",
 processor_id: nil,
 default: true,
 data: nil,
 deleted_at: nil,
 created_at: 2024-01-15 13:37:21.303918 UTC,
 updated_at: 2024-01-15 13:37:21.303918 UTC,
 stripe_account: nil,
 payment_method_token: nil>
```
Since both charges and subscriptions require the `customer_id for finding the costumer as:
https://github.com/pay-rails/pay/blob/main/lib/pay/paddle_billing/subscription.rb#L34
https://github.com/pay-rails/pay/blob/main/lib/pay/paddle_billing/charge.rb#L22

That will fail to update the record. Now comes the question: If I'm creating a new record, and `customer_id` is the Paddle customer ID, how come we require it at this point?

**Comment by deanpcmad** [comment, marked as answer]
In order for a Paddle Customer to be created, you need to run the `customer` method. Like so:

```ruby
# Set the payment processor
@user.set_payment_processor :paddle_billing

# Create the customer on Paddle
@user.payment_processor.customer
```

This creates the customer on Paddle and sets the `processor_id` on the `Pay::Customer` record

**Reply by fidalgo** [reply]
Thanks, that solved the problem!

**Comment by chrismahon** [comment]
Just wanted to add a quick follow up to this. I have the padding billing webhooks working as show in the documentation and I can see a successful subscription.

How can I get access to the payment method the customer is using though? I can't seem to find this information no matter what I try, and I can't see any entries into my `pay_payment_methods` table.

As everything for subscriptions is handled via webhooks with paddle billing I'm wondering if there is something I'm missing?

**Reply by deanpcmad** [reply]
It should set this on the `transaction.completed` webhook - https://github.com/pay-rails/pay/blob/main/lib/pay/paddle_billing/charge.rb#L53

**Reply by chrismahon** [reply]
thanks @deanpcmad does that webhook get triggered automatically or do I need to initiate it?

**Reply by deanpcmad** [reply]
If you've added that webhook on Paddle, it should trigger automatically. If you go to your [notifications](https://vendors.paddle.com/notifications) page, you can view the logs of what's been sent

**Reply by chrismahon** [reply]
thanks @deanpcmad yea I can see that webhook is setup on paddle

Maybe it's the way I have the subscription setup, but if I subscribe as a new customer the `pay_payment_methods` doesn't have anything for that customer  unless I update my payment method details after subscribing.

Once I do that I can see an entry gets added into the database, but not when I first subscribe. I have the subscription set to recur monthly and have not added any free trial period days, so my assumption was the first payment happens immediately?

Checking the logs I can see a `transaction.completed` went through successfully when I first subscribe.

I have sidekiq processing jobs locally and have my dev environment accessible using a Cloudflare Tunnel.

**Reply by deanpcmad** [reply]
Is a `Pay::Charge` being created too? Because it *should* create a payment method at the same time

**Reply by chrismahon** [reply]
hi @deanpcmad I can see an entry has gone into `pay_charges` for that subscription and that is also reflected on the Paddle dashboard as I can see the transaction / receipt etc.

**Reply by chrismahon** [reply]
hi @deanpcmad is there a way to manually trigger the `transaction.completed` webhook or check it's being fired off on my local logs? As mentioned the `Pay::Charge` seems to be happening as I can see something in the `pay_charges` table.

**Reply by deanpcmad** [reply]
Currently there's no way to sync the payment method, but I can work on that and create a PR.

**Reply by deanpcmad** [reply]
Just found this was a bug and I've created the PR #946. It also adds a `sync_from_transaction` method for pulling in the payment method details

**Reply by chrismahon** [reply]
Thanks @deanpcmad really appreciate that!

**Reply by chrismahon** [reply]
hi @deanpcmad, I've got payment methods being added to my database now after your PR which is great, thanks for sorting that bug out.

When I update the payment method I can see it adds another entry into the `pay_payment_methods` table but it doesn't update the customers `default_payment_method` when I try and access it like this:

```
<%= @user.payment_processor.default_payment_method %>
```

To get the last payment method I have to do this:

```
<%= @user.payment_processor.payment_methods.last %>
```

Is this the intended behaviour? If so, is there a way for me to use the paddle billing webhooks so that the last payment method added becomes the default?

**Reply by deanpcmad** [reply]
Glad to hear that's fixed it. That update should also make the payment method the default, but you can use `make_default!` to force a payment method to be default

**Reply by chrismahon** [reply]
hi @deanpcmad I'm definitely not seeing the default card being updated when I add a new one. I can see in my database there are 3 cards for the same person and they all have `true` set to the `default` column

<table>
	<tr>
		<th>id</th>
		<th>customer_id</th>
		<th>processor_id</th>
		<th>default</th>
		<th>type</th>
		<th>data</th>
		<th>created_at</th>
		<th>updated_at</th>
	</tr>
	<tr>
		<td>4</td>
		<td>22</td>
		<td>49f8675b-c658-4382-81a8-23e6fb1456ce</td>
		<td>True</td>
		<td>card</td>
		<td>{"brand": "visa", "last4": "4242", "exp_year": 2033, "exp_month": 12}</td>
		<td>2024-01-26 17:34:51.803763</td>
		<td>2024-01-26 17:34:51.803763</td>
	</tr>
	<tr>
		<td>5</td>
		<td>22</td>
		<td>0f3b0ab6-4b75-4f45-aced-f817ab37c77f</td>
		<td>True</td>
		<td>card</td>
		<td>{"brand": "visa", "last4": "5556", "exp_year": 2032, "exp_month": 12}</td>
		<td>2024-01-26 17:43:32.691047</td>
		<td>2024-01-26 17:43:32.691047</td>
	</tr>
	<tr>
		<td>6</td>
		<td>22</td>
		<td>0b621fb5-9d41-4a20-ab3d-b8bbd40ea09a</td>
		<td>True</td>
		<td>card</td>
		<td>{"brand": "visa", "last4": "4242", "exp_year": 2034, "exp_month": 10}</td>
		<td>2024-01-26 17:48:11.694551</td>
		<td>2024-01-26 17:48:11.694551</td>
	</tr>
</table>

**Reply by deanpcmad** [reply]
Oh oops. Have removed that from the PR

**Reply by chrismahon** [reply]
ah awesome thanks @deanpcmad really appreciate your help on this!
